searchEngine/SearchEngine.py

```diff
@@ -0,0 +1,55 @@
+from MockArticle import Article
+
+class SearchEngine:
+    """
+    This class implements the functions to score every article 
+    by the amount of incidences between keywords and synonyms to articles
+
+    Attributes:
+        keywords: A list of keywords
+        synonyms: A list of synonyms for the keywords
+        articles: A list of articles
+        keywords_weight: The weight that a keyword has on the scorig algorithm
+            Optional: default argument as 1
+        synonyms_weight: The weight that a synonym has on the scoring algorithm
+            Optional: default argument as 1
+    """
+    def __init__(self, articles):
+        """
+        Inits the search engine by loading all articles
+        """
+        self.articles = articles
+
+    def query(self, keywords, synonyms, keywords_weight=1, synonyms_weight=1):
+        """
+        Loads all necessary data from the query.
+        The weights attributes are default arguments for ease of use
+        """
+        self.keywords = keywords
+        self.synonyms = synonyms
+        self.keywords_weight = keywords_weight
+        self.synonyms_weight = synonyms_weight
+
+    def score_articles(self):
+        """
+        Counts the number of incidences between article words, keywords, and synonyms
+        implementing the different weights for the different type of weight values
+
+        Returns:
+            A map of the score for every article
+        """
+
+        score_per_article = {}
+        for article in self.articles:
+            score = 0
+            article_frequency = article.keywords
+            for keyword in self.keywords:
+                score += article_frequency.get(keyword, 0) * self.keywords_weight
+
+            for synonym in self.synonyms:
+                score += article_frequency.get(synonym, 0) * self.synonyms_weight
+
+            if score != 0:
+                score_per_article[article.id] = score
```

Comment:
If I understand correctly how this algorithm works, it seems that words which have more synonyms will always have more weight.

For example, if word A has only one available synonym, and word B has five available synonyms, the effective significance of word B will be 5x that of word A.

This is important for proper nouns, which are unlikely to have synonyms but (at least intuitively) are probably far more relevant to the search results.

Reply:
Yeah, I hadn't thought about that. Seems like a real issue. Maybe the synonym weight attribute can change depending on the amount of synonyms so that the effective significance is always the same, but that seems more like a patch than a design decision.

Reply:
Probably fine to leave as-is for the prototype and see how it performs.